odiglet/pkg/ebpf/director.go

```diff
@@ -201,95 +245,92 @@ func (d *EbpfDirector[T]) Instrument(ctx context.Context, pid int, pod types.Nam
-	go func() {
-		select {
-		case <-loadedCtx.Done():
-			return
-		case <-loadedIndicator:
-			d.instrumentationStatusChan <- instrumentationStatus{
-				Healthy:       true,
-				Message:       "Successfully loaded eBPF probes to pod: " + pod.String(),
-				Workload:      *podWorkload,
-				Reason:        LoadedSuccessfully,
-				PodName:       pod,
-				ContainerName: containerName,
-				Pid:           pid,
+	ip.runOnce.Do(func() {
+		loadedIndicator := make(chan struct{})
+		loadedCtx, loadedObserverCancel := context.WithCancel(ctx)
+		// launch an observer for successful loading of the eBPF probes
+		go func() {
+			select {
+			case <-loadedCtx.Done():
+				return
+			case <-loadedIndicator:
+				d.instrumentationStatusChan <- instrumentationStatus{
+					Healthy:       true,
+					Message:       "Successfully loaded eBPF probes to pod: " + pod.String(),
+					Workload:      *podWorkload,
+					Reason:        LoadedSuccessfully,
+					PodName:       pod,
+					ContainerName: containerName,
+					Pid:           pid,
+				}
 			}
-		}
-	}()
+		}()
 
-	go func() {
-		// once the instrumentation finished running (either by error or successful exit), we can cancel the 'loaded' observer for this instrumentation
-		defer loadedObserverCancel()
-		inst, err := d.instrumentationFactory.CreateEbpfInstrumentation(ctx, pid, appName, podWorkload, containerName, pod.Name, loadedIndicator)
-		if err != nil {
-			d.instrumentationStatusChan <- instrumentationStatus{
-				Healthy:       false,
-				Message:       err.Error(),
-				Workload:      *podWorkload,
-				Reason:        FailedToInitialize,
-				PodName:       pod,
-				ContainerName: containerName,
-				Pid:           pid,
+		go func() {
+			// once the instrumentation finished running (either by error or successful exit), we can cancel the 'loaded' observer for this instrumentation
+			defer loadedObserverCancel()
+			inst, err := d.instrumentationFactory.CreateEbpfInstrumentation(ctx, pid, appName, podWorkload, containerName, pod.Name, loadedIndicator)
+			if err != nil {
+				d.instrumentationStatusChan <- instrumentationStatus{
+					Healthy:       false,
+					Message:       err.Error(),
+					Workload:      *podWorkload,
+					Reason:        FailedToInitialize,
+					PodName:       pod,
+					ContainerName: containerName,
+					Pid:           pid,
+				}
+				return
 			}
-			return
-		}
 
-		d.mux.Lock()
-		_, stillExists := d.pidsAttemptedInstrumentation[pid]
-		if stillExists {
-			d.pidsToInstrumentation[pid] = inst
-			d.mux.Unlock()
-		} else {
-			d.mux.Unlock()
-			// we attempted to instrument this process, but it was already cleaned up
-			// so we need to clean up the instrumentation we just created
-			err = inst.Close(ctx)
-			if err != nil {
-				log.Logger.Error(err, "error cleaning up instrumentation for process", "pid", pid)
+			if ip.closed.Load() {
+				log.Logger.Info("Instrumentation already closed before running, stopping instrumentation", "pid", pid)
+				return
```

Comment:
above we called `d.instrumentationFactory.CreateEbpfInstrumentation` and here we return without cleaning up anything. Just making sure that the ebpf instrumentation does not hold any resources after creation

Reply:
If it is closed, it means the cleanup function was called for that instrumented process